### PR TITLE
Adjusted design for ShapeEnv.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97164
* #96699

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -338,13 +344,98 @@ def eval_guards(gm, *args):
...
+    there are NO guards as this is brittle because compilers should be able to
+    add extra constraints.  If you want to assert that there are no guards,
+    use StrictMinMaxConstraint with an unbounded ValueRanges.
+    """
+    pass
+
+# NB: None here indicates the client constraint is whatever is implicitly
+# inferred by guards from tracing, and that a backend can add whatever guards
+# it wants (including fully specializing the value).
+DimConstraint = Union[StrictMinMaxConstraint, RelaxedUnspecConstraint, None]
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:438]: I revamped DimConstraint hierarchy, check comments for what the constraints mean

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1424,36 +1509,59 @@ def create_unbacked_symint(self):
...
+        val: int,
+        source: Source,
+        dynamic_dim: DimDynamic,
+        constraint_dim: DimConstraint,
+    ) -> "sympy.Expr":
         assert isinstance(source, Source), f"{type(source)} {source}"
-        if dim_state == DimDynamismState.STATIC:
-            return sympy.Integer(val)
 
-        duck = dim_state == DimDynamismState.DYNAMIC
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1435]: You forgot to invert the meaning of duck

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1424,36 +1509,59 @@ def create_unbacked_symint(self):
...
-            self.var_to_range[sympy_expr] = _default_value_range(specialize_zero_one=self.specialize_zero_one)
+            # Apply default range, which assumes not zero-one
+            lower = 2 if self.specialize_zero_one else 0
+            self.var_to_range[sympy_expr] = ValueRanges(lower, sys.maxsize - 1)
+
+            # Small performance optimization: if we have a min-max constraint,
+            # we can proactively narrow to that range
+            if isinstance(constraint_dim, StrictMinMaxConstraint):
+                assert not duck
+                self.var_to_range[sympy_expr] &= constraint_dim.vr
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1562]: My plan is to NOT use var_to_range to check the constraints

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -204,37 +207,38 @@ def guard_scalar(a):
         raise AssertionError(f"unrecognized scalar {a}")
 
 # inclusive both ways
-def constrain_range(a, *, min: Optional[int], max: Optional[int] = None, user_directive=False):
+def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:210]: No longer calling constrain_range from constraints, there's a different strategy now.

**Comment by vanbasten23** [comment]
I wonder if there is a design doc for the adjusted ShapeEnv design

**Comment by voznesenskym** [comment]
Stilllll squashing tests here

**Comment by ezyang** [comment]
folded into parent pr

**Pull request closed without merging**